File: Hash Map - Set/2215 - FindTheDifferenceOfTwoArrays.py
"""
2215. Find the Difference of Two Arrays.

Given two 0-indexed integer arrays nums1 and nums2, return a list answer of size 2 where:

answer[0] is a list of all distinct integers in nums1 which are not present in nums2.
answer[1] is a list of all distinct integers in nums2 which are not present in nums1.
Note that the integers in the lists may be returned in any order.
"""

class Solution(object):
    def findDifference(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[List[int]]
        """
        answer1 = set()
        answer2 = set()

        set_nums1 = set(nums1)
        set_nums2 = set(nums2)

        for num in set_nums1:
            if num not in set_nums2:
                answer1.add(num)

        for num in set_nums2:
            if num not in set_nums1:
                answer2.add(num)

        return [list(answer1), list(answer2)]

s = Solution()

# Case 1
result1 = s.findDifference(nums1=[1,2,3], nums2=[2,4,6])
expected1 = [[1,3],[4,6]] 
print(f"Result: {result1}, Expected: {expected1}")
assert result1 == expected1

# Case 2
result2 = s.findDifference(nums1=[1,2,3,3], nums2=[1,1,2,2])
expected2 = [[3],[]] 
print(f"Result: {result2}, Expected: {expected2}")
assert result2 == expected2

File: HashMap&Set/2215_find_the_difference_of_two_arrays.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

class Solution{
    public:
    std::vector<std::vector<int>> findDifference(std::vector<int>& nums1, std::vector<int>& nums2){
        std::set<int> answer1;
        std::set<int> answer2;

        std::set<int> set_nums1(nums1.begin(), nums1.end());
        std::set<int> set_nums2(nums2.begin(), nums2.end());

        for (int num : set_nums1){
            if (set_nums2.find(num) == set_nums2.end()){
                answer1.insert(num);
            }
        }
        for (int num : set_nums2){
            if (set_nums1.find(num) == set_nums1.end()){
                answer2.insert(num);
            }
        }

        return {std::vector<int>(answer1.begin(), answer1.end()), std::vector<int>(answer2.begin(), answer2.end())};
    }
};

int main()
{
    Solution s;

    // Case 1
    std::vector<int> nums1_case1 = {1, 2, 3};
    std::vector<int> nums2_case1 = {2, 4, 6};
    std::vector<std::vector<int>> result1 = s.findDifference(nums1_case1, nums2_case1);
    std::vector<std::vector<int>> expected1 = {{1, 3}, {4, 6}};
    std::cout << "Result: {{" << result1[0][0] << "," << result1[0][1] << "},{" << result1[1][0] << "," << result1[1][1] << "}}, Expected: {{" << expected1[0][0] << "," << expected1[0][1] << "},{" << expected1[1][0] << "," << expected1[1][1] << "}}\n";

    std::vector<int> nums1_case2 = {1, 2, 3, 3};
    std::vector<int> nums2_case2 = {1, 1, 2, 2};
    std::vector<std::vector<int>> result2 = s.findDifference(nums1_case2, nums2_case2);
    std::vector<std::vector<int>> expected2 = {{3}, {}};
    std::cout << "Result: {{" << result2[0][0] << "},{}}, Expected: {{" << expected2[0][0] << "},{}}\n";

    return 0;
}